storage/database/sharded_database.go

```diff
@@ -151,102 +153,188 @@ func (pdb *shardedDB) Close() {
-		iterators = append(iterators, pdb.shards[i].NewIteratorWithStart(start))
+// NewIteratorWithStart creates a iterator over a subset of database content
+// starting at a particular initial key (or after, if it does not exist).
+func (db *shardedDB) NewIteratorWithStart(start []byte) Iterator {
+	return db.newIterator(func(db Database) Iterator { return db.NewIteratorWithStart(start) })
+}
+
+// NewIteratorWithPrefix creates a iterator over a subset of database content
+// with a particular key prefix.
+func (db *shardedDB) NewIteratorWithPrefix(prefix []byte) Iterator {
+	return db.newIterator(func(db Database) Iterator { return db.NewIteratorWithPrefix(prefix) })
+}
+
+func (db *shardedDB) newIterator(newIterator func(Database) Iterator) Iterator {
+	it := &shardedDBIterator{
+		db.NewChanIterator(context.Background(), newIterator),
+		make(chan entry, shardedDBCombineChanSize),
+		nil, nil}
+
+	go it.newCombineWorker()
+
+	return it
+}
+
+func (it shardedDBIterator) newCombineWorker() {
+	// TODO Put all items in biary-alphabetical order.
+	for {
+	chanIter:
+		for i, ch := range it.resultChs {
+			select {
+			case <-it.ctx.Done():
+				logger.Trace("[shardedDBIterator] combine worker ends due to ctx")
+				close(it.resultCh)
+				return
+			case e, ok := <-ch:
+				if !ok {
+					it.resultChs = append(it.resultChs[:i], it.resultChs[i+1:]...)
+					break chanIter
+				}
+				it.resultCh <- e
+			default:
+			}
+		}
+		if len(it.resultChs) == 0 {
+			logger.Trace("[shardedDBIterator] combine worker finishes iterating")
+			close(it.resultCh)
+			return
+		}
 	}
+}
 
-	for _, iter := range iterators {
-		if iter != nil {
-			if !iter.Next() {
-				iter = nil
+// Next gets the next item from iterators.
+// The first call of Next() might take 50ms.
+func (it *shardedDBIterator) Next() bool {
+	maxRetry := 5
+	for i := 0; i < maxRetry; i++ {
+		select {
+		case e, ok := <-it.resultCh:
+			if ok {
+				it.key, it.value = e.key, e.val
+				return true
+			} else {
+				logger.Error("[shardedDBIterator] Next is called on closed channel")
+				return false
 			}
+		default:
+			logger.Debug("[shardedDBIterator] no value is ready")
+			// shardedDBIterator needs some time for workers to fill channels
+			// If there is no sleep here, there should a sleep after NewIterator()
+			time.Sleep(10 * time.Millisecond)
+		}
+	}
+	logger.Error("[shardedDBIterator] Next() takes more than 50ms on unclosed subchannels")
+	return false
+}
+
+func (it *shardedDBIterator) Error() error {
+	var err error
+	for i, iter := range it.iterators {
+		if iter.Error() != nil {
+			logger.Error("[shardedDBIterator] error from iterator",
+				"err", err, "shardNum", i, "key", it.key, "val", it.value)
+			err = iter.Error()
+		}
+	}
+	return err
+}
+
+func (it *shardedDBIterator) Key() []byte {
+	return it.key
+}
+
+func (it *shardedDBIterator) Value() []byte {
+	return it.value
+}
+
+func (it *shardedDBIterator) Release() {
+	it.ctx.Done()
```

Comment:
I cannot understand this line. As far as I know, it cannot send a message to `ctx.Done()` channel. Could you explain what does it for?

Reply:
Thanks for pointing it out. It wasn't working properly.
I added cancel function to end context and test functions to test context.